test/unit_test/feature_group_test.cc

```diff
@@ -39,34 +43,85 @@ BOOST_AUTO_TEST_CASE(unique_feature_group_test)
   unique_features(fs, 5);
   check_collections_exact(
       std::vector<feature_index>(fs.indicies.begin(), fs.indicies.end()), std::vector<feature_index>{1, 2, 3, 5, 7});
+  check_collections_exact(fs.namespace_extents, std::vector<VW::namespace_extent>{{2, 4, 1}});
 
   // Uncapped
   unique_features(fs_copy1);
   check_collections_exact(std::vector<feature_index>(fs_copy1.indicies.begin(), fs_copy1.indicies.end()),
       std::vector<feature_index>{1, 2, 3, 5, 7, 11, 12, 13, 25});
+  check_collections_exact(fs_copy1.namespace_extents, std::vector<VW::namespace_extent>{{2, 4, 1}, {5, 8, 2}});
 
   // Special case at max 1
   unique_features(fs_copy2, 1);
   check_collections_exact(
       std::vector<feature_index>(fs_copy2.indicies.begin(), fs_copy2.indicies.end()), std::vector<feature_index>{1});
+  BOOST_REQUIRE(fs_copy2.namespace_extents.empty());
 
   // Special case for max 0
   unique_features(fs_copy3, 0);
   BOOST_REQUIRE(fs_copy3.empty());
+  BOOST_REQUIRE(fs_copy3.namespace_extents.empty());
 
   // Explicit negative input that isn't -1
   unique_features(fs_copy4, -10);
   check_collections_exact(std::vector<feature_index>(fs_copy4.indicies.begin(), fs_copy4.indicies.end()),
       std::vector<feature_index>{1, 2, 3, 5, 7, 11, 12, 13, 25});
+  check_collections_exact(fs_copy4.namespace_extents, std::vector<VW::namespace_extent>{{2, 4, 1}, {5, 8, 2}});
 
   // Special case for max 0
   features empty_features;
   unique_features(empty_features, 0);
   BOOST_REQUIRE(empty_features.empty());
+  BOOST_REQUIRE(empty_features.namespace_extents.empty());
 
   features fs_size_one;
   fs_size_one.push_back(1.f, 1);
   unique_features(fs_size_one);
   check_collections_exact(std::vector<feature_index>(fs_size_one.indicies.begin(), fs_size_one.indicies.end()),
       std::vector<feature_index>{1});
+  BOOST_REQUIRE(fs_size_one.namespace_extents.empty());
+}
+
+BOOST_AUTO_TEST_CASE(flatten_then_unflatten_namespace_extents_test)
+{
+  std::vector<VW::namespace_extent> extents{{0, 1, 1}, {1, 2, 2}};
+
+  auto flat_list = VW::details::flatten_namespace_extents(extents, 2);
+  auto unflattened_list = VW::details::unflatten_namespace_extents(flat_list);
+  BOOST_REQUIRE(unflattened_list == extents);
+
+  extents = {{1, 2, 123}};
+
+  flat_list = VW::details::flatten_namespace_extents(extents, 2);
+  unflattened_list = VW::details::unflatten_namespace_extents(flat_list);
+  BOOST_REQUIRE(unflattened_list == extents);
+
+  extents = {{1, 2, 1}, {5, 8, 2}, {8, 12, 3}, {13, 14, 4}};
+  flat_list = VW::details::flatten_namespace_extents(extents, 18);
```

Comment:
how are the 2's and 18 calculated from the extents?

Reply:
18 is the length of the feature_group. This is separate because there can be a range with no hash (I intend to make it required, but it isn't in this PR). In this example positions 14, 15, 16 and 17 do not have a hash since they are after the last extent but before the end at position 18.

An extent is a tuple of `(start_index, end_index, hash)` so this maps to `{1, 2, 1}`. The range is `[start_index, end_index)`